cmd/jaeger/internal/integration/span_reader.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package integration
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/ports"
+	"github.com/jaegertracing/jaeger/proto-gen/api_v2"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+var (
+	_ spanstore.Reader = (*spanReader)(nil)
+	_ io.Closer        = (*spanReader)(nil)
+)
+
+// SpanReader retrieve span data from Jaeger-v2 query with api_v2.QueryServiceClient.
+type spanReader struct {
+	Port int
+
+	clientConn *grpc.ClientConn
+	client     api_v2.QueryServiceClient
+}
+
+func createSpanReader(port int) *spanReader {
+	return &spanReader{
+		Port: port,
+	}
+}
+
+func (r *spanReader) Start() error {
+	opts := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	cc, err := grpc.DialContext(ctx, ports.PortToHostPort(ports.QueryGRPC), opts...)
+	if err != nil {
+		return err
+	}
+
+	r.clientConn = cc
+	r.client = api_v2.NewQueryServiceClient(cc)
+	return nil
+}
+
+func (r *spanReader) Close() error {
+	return r.clientConn.Close()
+}
+
+func unwrapNotFoundErr(err error) error {
+	if s, _ := status.FromError(err); s != nil {
+		if strings.Contains(s.Message(), spanstore.ErrTraceNotFound.Error()) {
+			return spanstore.ErrTraceNotFound
+		}
+	}
+	return err
+}
+
+func (r *spanReader) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error) {
```

Comment:
a bit sad that you had to implement all of that. This is a takeway for V2 storage - perhaps it's possible to have remote storage API to be the same as query service API. Then we could've used plugin/storage/grpc/shared/client

Reply:
Okay, I'll create a new issue for this task.